### DISCUSSION TITLE
Proxying another gRPC service

### DISCUSSION THREAD

**Discussion by bouk** [hyperium/tonic#1357] [Q&A]
Hi, I have a Python gRPC service that I want to proxy using tonic. I've seen this post: https://github.com/hyperium/tonic/discussions/1111 but I haven't been able to figure out how I can extend it to reverse proxy the request to another service.

Specifically I'd like to use tonic-web to add grpc web support for my Python grpc service. I know there's other software that can do this but we already have a rust service and it'd be nice not to introduce another component. Does anyone have experience with this?

**Comment by drauschenbach** [comment]
The challenge here is to forward a `request.metadata()` metadata map into a client intercept function, but those intercept functions are function pointers and so they don't allow closures that do any variable capture.

**Comment by drauschenbach** [comment]
@yotamofek because of your work on #713, do you have any input here?

**Comment by drauschenbach** [comment]
The solution is to write an interceptor struct, like [in this example](https://github.com/hyperium/tonic/blob/master/examples/src/interceptor/client.rs).

```rust
struct MyAuthorizer {
    authorization: MetadataValue<Ascii>,
}

impl Interceptor for MyAuthorizer {
    fn call(&mut self, mut req: Request<()>) -> Result<Request<()>, Status> {
        req.metadata_mut().insert("authorization", self.authorization.clone());
        Ok(req)
    }
}

let authorization = req.metadata().get("authorization").clone().unwrap().clone();
let authorizer = MyAuthorizer { authorization };
let mut my_client = MyClient::with_interceptor(channel.clone(), authorizer);
```
